### PR TITLE
[dynamo] add config to report guard failure values

### PR DESCRIPTION
**Pull request by jon-chuang**
Follow up to https://github.com/pytorch/pytorch/pull/110927

Example output:

With `report_guard_failure_values=True`.

With all failures enabled:
```
[2023-10-10 02:04:37,366] torch._dynamo.convert_frame.__recompiles: [DEBUG] ('Recompiling function inner in /home/jonch/Desktop/Programming/mlsys/pytorch/torch/_dynamo/external_utils.py:15', 'triggered by the following guard failures: [\'___guarded_code.valid\', "___check_obj_id(L[\'args\'][1], 140039813517312), with LHS value: {\'id\': 140040114487792}", "___check_obj_id(L[\'args\'][2], 140039813517168), with LHS value: {\'id\': 140040124395392}"]')

```

With default first failure:

```
.[2023-10-10 01:56:07,749] torch._dynamo.convert_frame.__recompiles: [DEBUG] ('Recompiling function fn in /home/jonch/Desktop/Programming/mlsys/pytorch/test/dynamo/test_misc.py:1354', "triggered by the following guard failure: L['a'] == 0, with LHS value: 1")
[2023-10-10 01:56:07,757] torch._dynamo.convert_frame.__recompiles: [DEBUG] ('Recompiling function fn in /home/jonch/Desktop/Programming/mlsys/pytorch/test/dynamo/test_misc.py:1354', "triggered by the following guard failure: tensor '__as_tensor(L['i'])' size mismatch at index 0. expected 1, actual 5")
[2023-10-10 01:56:07,767] torch._dynamo.convert_frame.__recompiles: [DEBUG] ('Recompiling function fn in /home/jonch/Desktop/Programming/mlsys/pytorch/test/dynamo/test_misc.py:1354', "triggered by the following guard failure: ___check_type_id(L['a'], 94852103176992), with LHS value: {'type': <class 'NoneType'>}")
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @lezcano @eellison

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -1347,6 +1371,11 @@ def guard_fail_hook(
         if isinstance(fail_reason, bool) and not fail_reason:
             fail_reason = part
         if isinstance(fail_reason, str):
+            if config.report_guard_failure_values:
+                if lhs := get_lhs_val(part):
+                    lhs_value = eval(lhs, global_scope, scope)
+                    fail_reason += f", with LHS value: {lhs_value}"
```
Comment (lezcano) [inline torch/_dynamo/guards.py:1377]: what about
```suggestion
                    fail_reason += f", when executed with value: {lhs_value}"
```

Comment (jon-chuang) [reply torch/_dynamo/guards.py:1377]: Hmm, it seemed clearer to me `with LHS value`. 

I wouldn't know which "executed value" the logs referred to.

See e.g.

```
["___check_obj_id(L[\'args\'][1], 140039813517312), with LHS value: {\'id\': 140040114487792}", "___check_obj_id(L[\'args\'][2], 140039813517168), with LHS value: {\'id\': 140040124395392}"]
```

Comment (lezcano) [reply torch/_dynamo/guards.py:1377]: what about:
>"triggered by the following guard failure: Expected L['a'] == 0, but got L['a'] ==  1"

Comment (jon-chuang) [reply torch/_dynamo/guards.py:1377]: Hmm sounds good. It's definitely clearer (this was just a hacky way of getting some useful output)

**Comment by ezyang** [comment]
This is a lot of code, with no tests. It looks fairly tricky, with evals and regexes. And it is hugely undiscoverable, not hooked into the default logging system.

Here's my challenge to you: can you figure out a way to report this information **by default**, without having to toggle on a configuration? This will give us confidence that it works in all situations, because you will be exercising the code for all runs. It will require you to work out how to only run the guard failure reporting logic when you actually trigger a compile, because it is not acceptable to run this logic whenever a cache entry fails guards (there might be a hit somewhere else.)

It is also a little awkward because @anijain2305 is working on a refactor of the guard infrastructure to move it into C++, which is also likely to change what the best way to go about doing this is.

cc @jansel

**Comment by jon-chuang** [comment]
> It is also a little awkward because @anijain2305 is working on a refactor of the guard infrastructure to move it into C++, which is also likely to change what the best way to go about doing this is.

Sure, I can hold off this change until the guard code has stabilized. Regarding this change, I suppose the motivation is perf? Moving into C++ would definitely make it harded to define new guards.

**Comment by ezyang** [comment]
> Regarding the refactor, I suppose the motivation is perf? Moving into C++ would definitely make it harder to define new guards.

yep

> This doesn't seem too hard, we can just accumulate the check_fns and lazily evaluate them on recompile. We can also better structure the eval/guard function definitions so as to only run eval once (we can have an alternate check function that accumulates all the intermediate results into a dict, else use an environment dict to get the output).

You don't even need to accumulate them; @anijain2305 made it so you can get at the cache entries from Python, so you can literally just rerun all the guards again in the compiler.

**Comment by jon-chuang** [comment]
Freezing this until @anijain2305's work is complete.

**Comment by jon-chuang** [comment]
Let's continue this since we've officially completed moving the guards failure reporting off the non debug-mode path

**Comment by ezyang** [comment]
Rereviewing this, I think my preference would be for us to separately store an "interesting expression" which you can evaluate to get whatever it is you want to print, instead of extracting it via splitting strings on == which ugh ugh ugh

**Comment by jon-chuang** [comment]
> instead of extracting it via splitting strings on == which ugh ugh ugh

Yes, we can do so via similar "extras" mechanism to verbose_code_parts, maybe a extra formatting of the return value at the end of verbose_code_parts

**Pull request closed without merging**